File: 2020/0416-partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        # 基础检查
        if not nums:
            return False
        n = len(nums)
        if n < 2:
            return False
        
        total = sum(nums)
        target = total //2
        if total % 2 != 0:
            return False

        # 记忆化递归
        dp = [set() for _ in range(n)]
        def recursion(nums, i, remain):
            if remain == 0:
                return True
            # 剩余值小于0、数组遍历完了、已经计算过了并且还没退出返回则同样不要计算了
            if remain < 0 or i == n or remain in dp[i]:
                return False
            dp[i].add(remain)
            #  选择当前nums[i]或者不选
            return recursion(nums, i+1, remain- nums[i]) or recursion(nums, i+1, remain)

        return recursion(nums, 0, target)

File: hot100/0416-partition-equal-subset-sum.go
func canPartition(nums []int) bool {
    // 检查
    n := len(nums)
    if n < 2 {
        return false
    }

    sum, max := 0, 0
    for _, v := range nums {
        sum += v
        if v > max {
            max = v
        }
    }
    if sum%2 != 0 {
        return false
    }

    target := sum / 2
    if max > target {
        return false
    }

    // dp
    dp := make([]bool, target+1)
    // 初始化
    dp[0], dp[nums[0]] = true, true
    for i := 1; i < n; i++ {
        v := nums[i]
        // 为了保证dp[j] = dp[j] || dp[j-v] 能用到上一排的dp[j-v]，需要从大到小更新dp值
        for j := target; j >= v; j-- {
            dp[j] = dp[j] || dp[j-v]
        }
    }

    return dp[target]
}
